### PR TITLE
Peg isolates to a thread, replace all usage of tokio threaded runtime with tokio's basic runtime

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

File: cli/compilers/ts.rs [hunk 1]
```diff
@@ -625,6 +623,38 @@ impl TsCompiler {
...
+      let msg = (worker.get_message().await).unwrap();
+      let json_str = std::str::from_utf8(&msg).unwrap();
+      load_sender.send(Ok(json!(json_str))).unwrap();
+    };
+    crate::tokio_util::run_basic(fut);
+  });
+
+  let fut = async { load_receiver.await.unwrap() };
+  fut.boxed_local()
+}
```
Comment (bartlomieju) [inline cli/compilers/ts.rs]: 👍

File: cli/ops/process.rs [hunk 2]
```diff
@@ -214,14 +188,17 @@ fn op_run_status(
...
 
   let future = async move {
-    let run_status = future.await?;
+    let run_status = {
+      let mut table = state.lock_resource_table();
+      let child_resource = table
+        .get_mut::<ChildResource>(rid)
+        .ok_or_else(bad_resource)?;
+      (&mut child_resource.child).await.map_err(ErrBox::from)?
+    };
```
Comment (bartlomieju) [inline cli/ops/process.rs]: Holds resource table mutex until `await` resolves - that's bad

File: cli/ops/process.rs [hunk 3]
```diff
@@ -241,10 +218,11 @@ fn op_run_status(
     }))
   };
 
-  let pool = futures::executor::ThreadPool::new().unwrap();
-  let handle = pool.spawn_with_handle(future).unwrap();
+  // TODO(ry) I don't understand why this needs to be in a pool?
+  //let pool = futures::executor::ThreadPool::new().unwrap();
+  //let handle = pool.spawn_with_handle(future).unwrap();
```
Comment (bartlomieju) [inline cli/ops/process.rs]: I think this was some legacy code that was never fixed 😢

File: cli/state.rs [hunk 4]
```diff
@@ -314,10 +319,13 @@ impl ThreadSafeState {
...
-  pub fn add_child_worker(&self, worker: WebWorker) -> u32 {
+  pub fn add_child_worker(&self, _worker: &WebWorker) -> u32 {
     let worker_id = self.next_worker_id.fetch_add(1, Ordering::Relaxed) as u32;
-    let mut workers_tl = self.workers.lock().unwrap();
-    workers_tl.insert(worker_id, worker);
+    println!("WARNING broken add_child_worker {}", worker_id);
+    // TODO(ry) Fix this... we can't have owned references to workers. so what
+    // goes in the table?
+    // let workers_tl = self.workers.lock().unwrap();
+    // workers_tl.insert(worker_id, worker);
```
Comment (bartlomieju) [inline cli/state.rs]: A few months ago we had something called "shared future" that was returned from worker and it was stored in the table - I don't think that's a good idea though, host worker needs to have "full" control over child worker (like termination, emptying pending futures, etc)

File: cli/ops/process.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/ops/process.rs]: @ry this statement still hold true, I suggest putting `child_resource.child` behind a mutex and cloning it - that way resource table won't be locked.

File: cli/ops/process.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/ops/process.rs]: Remove this TODO?

File: cli/lib.rs [hunk 5]
```diff
@@ -276,8 +277,8 @@ async fn info_command(flags: DenoFlags) {
   if let Err(e) = main_result {
     print_err_and_exit(e);
   }
-  print_file_info(worker.clone(), main_module.clone()).await;
-  let result = worker.await;
+  print_file_info(&worker, main_module.clone()).await;
+  let result = (&mut *worker).await;
```
Comment (bartlomieju) [inline cli/lib.rs]: Is there some syntactic sugar that could be used here?

File: cli/lib.rs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
-#![deny(warnings)]
+// #![deny(warnings)]
+#![allow(dead_code)]
```
Comment (bartlomieju) [inline cli/lib.rs]: Do you want to land it with this attribute?

File: cli/lib.rs (hunk 5 above)
Comment (ry) [reply cli/lib.rs]: *shrug* I'm not sure we should even have the worker be a future. I'd just leave it for now until we figure out the event loop.

File: cli/ops/process.rs (hunk 2 above)
Comment (ry) [reply cli/ops/process.rs]: I've undone all the changes to this module - turns out to be unrelated.

**Comment by piscisaureus** [comment]
So the Isolate is still a future if I'm reading this correctly.
Is there a way to make the V8 thread drive the tokio event loop instead of the other way around? Otherwise I don't see it going to help much with the Locker issue.

**Comment by piscisaureus** [comment]
2. Does this mean that all blocking ops (e.g. DNS resolution) are now actually blocking, or is there still a thread pool for that?

**Comment by ry** [comment]
> So the Isolate is still a future if I'm reading this correctly. Is there a way to make the V8 thread drive the tokio event loop instead of the other way around? Otherwise I don't see it going to help much with the Locker issue.

Probably. The event loop is basically defined the same as it was in `deno_core::Isolate::poll`. Now that they are all on their own threads we can start to dismantle that if we want to. I avoided making more changes than necessary to switch Tokio runtimes.

> Does this mean that all blocking ops (e.g. DNS resolution) are now actually blocking, or is there still a thread pool for that?

In this patch all blocking ops are done on their own thread. I leave it to future work to optimize this.
https://github.com/ry/deno/blob/b5ca5e6f0aa2cbb732827f745bb5e7cf15641a17/cli/ops/dispatch_json.rs#L106-L113

**Comment by piscisaureus** [comment]
> In this patch all blocking ops are done on their own thread. I leave it to future work to optimize this.
> https://github.com/ry/deno/blob/b5ca5e6f0aa2cbb732827f745bb5e7cf15641a17/cli/ops/dispatch_json.rs#L106-L113

Although the code looks fine from my perspective, I worry that it's going to hurt some benchmark performance badly. Do you have any numbers for e.g. module fetch / proxy performance?

File: cli/worker.rs [hunk 7]
```diff
@@ -45,10 +64,9 @@ pub struct WorkerChannels {
 ///  - `MainWorker`
 ///  - `CompilerWorker`
 ///  - `WebWorker`
-#[derive(Clone)]
 pub struct Worker {
   pub name: String,
-  pub isolate: Arc<AsyncMutex<Box<deno_core::EsIsolate>>>,
```
Comment (ry) [inline cli/worker.rs]: Note that we've gotten rid of this Arc Mutex

**Comment by ry** [comment]
> Although the code looks fine from my perspective, I worry that it's going to hurt some benchmark performance badly. Do you have any numbers for e.g. module fetch / proxy performance?

I don't think we measure any blocking ops in the benchmarks. Regarding compiling - we spawn a new thread for the TS compiler worker - so that's an async op.

**Comment by piscisaureus** [comment]
> In this patch all blocking ops are done on their own thread. I leave it to future work to optimize this.

I'm a bit afraid this is is never going to happen.

**Comment by hustxiaoc** [comment]
Are we going to use `threaded_scheduler`  runtime someday for handling  async ops  with multiple threading?

**Comment by ry** [comment]
@hustxiaoc Unlikely - we have in the past and have found the single threaded runtime faster and simpler for our purposes. I'd be happy to describe this in more detail if you're interested...

**Comment by hustxiaoc** [comment]
Sure, I'm happy to learn more detail about this :)

**Pull request merged**